### ISSUE TITLE
Design Thread-Safe Library

### ISSUE THREAD

**Issue by bwbeach** [Backblaze/B2_Command_Line_Tool#78]
We have a goal of making the B2 library and command-line tool thread safe.  Users of the library (both B2RawApi and B2Api) will want to multi-thread their applications.  And we want to multi-thread the command-line tool so that it can upload and download many files in parallel.

This issue is a place to discuss the design issues for the library.

**Comment by bwbeach** [comment]
The first question I'd like to address is what threading model to use.  

I think that the standard `threading` library will make the B2 library the most accessible.  If we were to code with `greenlet`, then users of the library would have to use `greenlet`.  But if we just use `threading`, then users can use standard threads, or they can use `greenlet`.

**Comment by ppolewicz** [comment]
I think if a `greenlet` takes a threading lock, it will block execution of all other greenlets (unless _monkeypatching_ of `Threading` module will be done, but the user doesn't have to do it, so we should not rely on it).

Instead, we can give the user an ability to provide a lock object factory (`threading.RLock` or `gevent.lock.RLock`, so that he can customize the way we synchronize with his code).

I don't think anyone should use `B2RawApi` - I was actually thinking of making it private (renaming to `_B2RawApi`).

**Comment by ppolewicz** [comment]
`StoredAccountInfo` should use locks to prevent corruption of the data file

**Comment by ppolewicz** [comment]
`portalocker` can be used for that as described here https://github.com/WoLpH/portalocker#examples

**Comment by ppolewicz** [comment]
alternative libraries
* `fasteners`: http://fasteners.readthedocs.org/en/latest/api/process_lock.html

`oslo.concurrency` (https://pypi.python.org/pypi/oslo.concurrency) seems to be an alternative, but it just wraps `fasteners.InterProcessLock`.

**Comment by bwbeach** [comment]
`urllib2` is not thread safe.  I've used [requests](https://github.com/kennethreitz/requests) before and like it.

**Comment by bwbeach** [comment]
What should the command-line tool report as it's doing parallel uploads?

I think that when uploading a single large file, it makes sense to put up a progress bar.  Updates to the progress bar will need to come from all of the threads uploading the parts of the large file.

For the sync command, which may upload several different files in parallel, things are more difficult.  One plan is to just report each file (or part of large file) when it finishes uploading.  This could mean waiting quite a while for an update, if you're uploading 100MB at a time.  An alternative is to put up a progress bar for the first file that starts, update it until it's done.  Then, once it's done, it can put up the progress bar for the next file, which may already be partly (or completely) done.  This would always show some indication of progress.

**Comment by ppolewicz** [comment]
A similar situation is described here: https://github.com/tqdm/tqdm/blob/master/README.rst#how-to-make-a-good-progress-bar

For multiple files the best we can do now is to report progress in total bytes. I can imagine reporting progress of each file as it is being uploaded (tqdm doesn't support it) but that would not be useful anyway. I think reporting total is better.

**Comment by bwbeach** [comment]
I kind of like the style of the AWS command-line tool.  It has a simple text progress line that counts files and parts.  As each file is completed, that is reported, and then the progress line is re-printed.

We could do something similar, but I think bytes are the important thing to count.

Our progress line could look something like this:

    scan: 463 of 10339 files   transfer: 23 files, 375.34 MB

And with printing when each file is completed, the output would like this:

    ...
    uploaded: b2://my-bucket/my-dir/file1.txt
    deleted: b2://my-bucket/my-dir/file2.txt
    scan: 463 of 10339 files   transfer: 23 files, 375.34 MB

**Comment by ppolewicz** [comment]
I think the essential information is:
- how fast is the transfer going
- how much longer will it take

and then "how much was already done" and a progress bar is a bonus.

We could do it like AWS does it, but personally, if I had a choice between what `tqdm` can provide for us (almost by default), I'd prefer our design. If we have a good idea on how to do this, then they should change their tool so that it looks more like ours does ;)

**Comment by bwbeach** [comment]
The question "how much longer will it take" is the difficult part.  I want the sync command to work on large file sets, and don't want it to have to fully compare the file lists (local and B2) before starting any transfers. 

How about three progress bars in sequence, with these titles:

* Step 1/3: count local files
* Step 2/3: compare file lists
* Step 3/3: transfer data

The first one should go fairly fast.  It won't know the total number of files, but you can watch the counter go up until it's done.  Then it will know the number of file names that need to be compared in step 2, which can have a real progress bar.  Step 2 will take a little longer because it has to fetch file lists from B2 and get the sizes of local files.  Then, when it's done comparing file lists, it can start reporting on the data transfer and know how much data will need to be transferred.

**Comment by ppolewicz** [comment]
I am not sure if tqdm can display 3 progress bars. I hope it can.

By the way, we should parallelize those 3 steps as much as possible. We could display progress very easily if they were sequential, but we can start sending files as soon as we have an authoritative answer of a file being missing from somewhere where it should be (for example we finish iterating over local directory and see that there are 5 files there, while b2 already answered that there are 6 files in the cloud in that directory - we should start downloading or deleting a file right now).

**Comment by bwbeach** [comment]
Yes, those tasks should definitely happen in parallel.

And I'm thinking three progress bars in sequence, one at a time.

**Comment by bwbeach** [comment]
Next topic: the StoredAccountInfo class.

The B2 service enforces a rule that each upload URL / auth token pair can be used for only one upload at a time.  This was a recent addition to the [B2 documentation](https://www.backblaze.com/b2/docs/uploading.html).

The StoredAccountInfo class needs to ensure that upload URLs are not shared among multiple processes running at the same time.  I propose that the interface to the class be changed to support a pool of upload URLs for each bucket, and that the threads doing uploading take one, use it, and then put it back in the pool.

**Comment by ppolewicz** [comment]
The fact that I am finding out about this restriction here is a concern for me. From my point of view b2interface has changed and I should adjust my code.

As for the design you suggest, I agree fully.

As for implementation, it seems that we need to start using sqllite for StoredAccountInfo

**Comment by bwbeach** [comment]
I agree that the API change was communicated poorly.  It went out in release notes, but was not clear enough.

I'm not sure about sqlite.  It may be more trouble that it's worth.
